scd/usbtoken.c

```diff
@@ -0,0 +1,737 @@
+/*
+ * This file is part of trust|me
+ * Copyright(c) 2013 - 2020 Fraunhofer AISEC
+ * Fraunhofer-Gesellschaft zur Förderung der angewandten Forschung e.V.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2 (GPL 2), as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GPL 2 license for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, see <http://www.gnu.org/licenses/>
+ *
+ * The full GNU General Public License is included in this distribution in
+ * the file called "COPYING".
+ *
+ * Contact Information:
+ * Fraunhofer AISEC <trustme@aisec.fraunhofer.de>
+ */
+#include "usbtoken.h"
+#include "ssl_util.h"
+
+#include "sc-hsm-cardservice.h"
+#include <ctapi.h>
+
+// #define LOGF_LOG_MIN_PRIO LOGF_PRIO_TRACE
+#include "common/macro.h"
+#include "common/mem.h"
+#include "common/file.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+
+#define TOKEN_MAX_AUTH_CODE_LEN 16
+#define TOKEN_KEY_LEN 32 /* must be coordinated with ssl_util.c */
```

Comment:
(TOKEN_MAX_AUTH_CODE_LEN != TOKEN_KEY_LEN) evaluates to true ;)

TOKEN_MAX_AUTH_CODE_LEN defines the number of bytes the usbtoken expects as authentication code.
TOKEN_KEY_LEN defines the number of key bytes we request the usbtoken to return to us for key wrapping

central definition is difficult and maybe not really beneficial without further checking valid combinations of the employed OpenSSL key wrapping cipher and the key length. these are sane values and should be fine. changes to them should only be made carefully

Reply:
i agree on that, please add a short comment to each of those defines indicating their use

Anyway, if the value "must be coordinated with ssl_util.c" it should be defined in the same place,
to avoid programming errors changing ssl_util.c without adjusting the TOKEN_KEY_LEN